### DISCUSSION TITLE
Do we really need the swap buttons for players in scenario selection?

### DISCUSSION THREAD

**Discussion by zenseii** [ihhub/fheroes2#4701] [Ideas]
Are these arrows for quickly moving name and class between two compatible player spots really necessary? They were added for fheroes2.
![image](https://user-images.githubusercontent.com/12501091/143595918-d4337c9b-8501-4eef-a1af-ae79f2d37cf6.png)

If the problem is that picking a class is a chore in the case you decide to play as a different color after having set it all up, then this can easily be solved by letting you use the scroll wheel to go back and forth between classes on the same color (which I think is a feature that should be implemented either way).

Also these arrows are flawed right now since they are present between spots that you cannot even swap between. If we decide to keep them then this should be addressed so that arrows will only be under spots that are swappable, taking into account the spots that have non-swappable spots between them.

**Comment by a1exsh** [comment]
>  use the scroll wheel to go back and forth between classes

Indeed. Should be easy to add — we did it recently for sound/music volume.

**Reply by zenseii** [reply]
Yes, I was actually hoping you'd notice this seeing that you've been busy adding this lately whenever you've not been working on the shadows.

**Reply by a1exsh** [reply]
https://github.com/ihhub/fheroes2/pull/4720 ;)

**Reply by zenseii** [reply]
Great job! Thanks!

**Comment by LeHerosInconnu** [comment]
Hello @zenseii and @a1exsh,

Originally, these arrows were added to change the order of play of the players during the turn (blue player, green player, etc.) in the old version of fheroes2 (the previous team).
For example, to make the game more difficult and make some AI players play first (so before the human player).
I have never done any test on this subject, neither in the old version, nor in the new one.

**Reply by zenseii** [reply]
Interesting and I think that is a cool feature. However, I believe a better implementation than those arrows could be made to change the order of who gets to do their turn when, like a simple number showing the turn order or something above the player icon, because simply swapping the colors doesn't explain the player anything. Having the AI act first gives room for some interesting map introductions, but I digress.

Basically what you're saying is that the main function of these arrows doesn't work right now anyways? Also their icons do look a lot lower resolution than the rest of this beautiful window, probably because they are upscales of the the cursor for meeting heroes, but that is a bit besides the point.

**Reply by LeHerosInconnu** [reply]
@zenseii.

> Interesting and I think that is a cool feature. However, I believe a better implementation than those arrows could be made to change the order of who gets to do their turn when, like a simple number showing the turn order or something above the player icon, because simply swapping the colors doesn't explain the player anything. Having the AI act first gives room for some interesting map introductions, but I digress.

In the original game, the human player always plays first.
If the human player chooses the orange color, the order is, human orange player, AI blue player, AI green player, etc.

In original game:

https://user-images.githubusercontent.com/43583994/143624715-db94b7b1-53a2-4620-a812-66b361bf70e1.mp4

With this feature, it becomes possible to change the order of play of the players.
If the user does not change the order of the players (by clicking on the red two-way arrows), then everything works as in the original game.
If the user changes the order of the players (by clicking on the red two-way arrows), then the players play from the left to the right, starting from the far left.

In old version of fheroes2 (previous team):

https://user-images.githubusercontent.com/43583994/143624684-bf5126b8-1e43-4679-8351-df7998815ddd.mp4

> Basically what you're saying is that the main function of these arrows doesn't work right now anyways? Also their icons do look a lot lower resolution than the rest of this beautiful window, probably because they are upscales of the the cursor for meeting heroes, but that is a bit besides the point.

The graphics are not great, but it was not the "final" version. :)
And there are also bugs in the old version of fheroes2 (previous team).
This no longer works in the latest version of fheroes2. :(

**Reply by zenseii** [reply]
Thanks for taking the effort to test this! I think this is a good feature to have indeed. What do you think about introducing the shift key  (again) or any key combined with the mouse as a way to change the order/place of the players? This way you could change the first and last player without having to click for every single step of swapping them.

**Reply by LeHerosInconnu** [reply]
@zenseii.

> Thanks for taking the effort to test this! I think this is a good feature to have indeed. What do you think about introducing the shift key (again) or any key combined with the mouse as a way to change the order/place of the players? This way you could change the first and last player without having to click for every single step of swapping them.

For my part, I prefer that it is possible to do everything with only the mouse (the same action with the use of the keyboard should be an alternative).

Why not use the "drag and drop" in this case?
It seems to be quite appropriate here.
In addition with the "drag and drop" it is possible to switch two players, or to insert a player between two others depending on the drop area. :)

**Reply by zenseii** [reply]
Yeah drag and drop would be perfect too. Excuse me for forgetting that people play exclusively with the mouse, since I'm the opposite 😅

The only reason I didn't mention drag and drop is because I don't know if there is code in place for this in the engine.

**Reply by LeHerosInconnu** [reply]
@zenseii.

> Yeah drag and drop would be perfect too. Excuse me for forgetting that people play exclusively with the mouse, since I'm the opposite :sweat_smile:
> 
> The only reason I didn't mention drag and drop is because I don't know if there is code in place for this in the engine.

There is something similar ("clic and move") with the "Fast separation into slots:" for the hero's army. :)

**Edit.**

This can also be done with the mouse wheel. :)
A little less flexible in use than the "drag and drop" but fast enough.

**Reply by zenseii** [reply]
> LeHerosInconnu

How would the mouse wheel implementation work? Would it just scroll every player after the other in a constant sequence or did you have something else in mind? I have an alternative solution explained below using a sequence of left and right mouse button.

**Reply by LeHerosInconnu** [reply]
Hello @zenseii,

> > LeHerosInconnu
> 
> How would the mouse wheel implementation work? Would it just scroll every player after the other in a constant sequence or did you have something else in mind? I have an alternative solution explained below using a sequence of left and right mouse button.

How to correctly interpret the sequence with the left and right buttons?
The user can left-click on a player for no particular reason, and then right-click on another player (without intending to swap them) because the right-click also displays an "Opponents" information window.

In fheroes2:

![Right clic opponent info window 01](https://user-images.githubusercontent.com/43583994/143718602-042a038b-4ae8-4b15-bf5f-190571d6d602.png)

And also, how to differentiate this from player selection.
Let's say that in the scenario window, the Blue player is selected by default.
![Human player's color selection 001](https://user-images.githubusercontent.com/43583994/143718704-49cc78a2-58bc-41c3-ab13-51b45ec48c2c.png)

Now, if the user wants to play another player (another color), he left-clicks for example on the orange player's graphic and this one is selected (as human player).
![Human player's color selection 002](https://user-images.githubusercontent.com/43583994/143718713-a2577c2e-63eb-4f6b-aaf0-2d1096a0f53a.png)

So the user should make a second click on the same graphic and a selection frame should be displayed around the player's graphic to indicate to the user that the interface has switched to a player order change mode and that the next click on another player's graphic will swap their positions in the play order.
The problem is that the first click on a player's graphic will change that player to a human player if he wasn't before (if it was possible to play that color as a human player).
And then the user will have to reselect the color he wants to play as a human player.
Which is a bit confusing and not very convenient.

For the mouse wheel operation, the user positions the mouse cursor on a player's graphic without clicking.
By moving the mouse wheel downwards, this player is moved one step to the right.
By moving the mouse wheel up, this player is moved one step to the left.
If the player is placed at the far left or far right (at the edge, without a player on each side), the player is also moved to the other extremity (the players are considered as in a "loop").
As long as the user does not move the mouse cursor outside the graphic of the original player whose position he has just changed, it is always this original player who changes his position if the user continues to use the mouse wheel in one direction or the other.
If the user wants to change the position of another player, he should move the mouse cursor over the graphic of this other player first.

For example, with the Blue player graphic (far left) as the origin, the standard order is:

Blue, Green, Red, Yellow, Orange, Purple.

With a click of the mouse wheel downwards, the order becomes:

Green, Blue, Red, Yellow, Orange, Purple.

With another click of the mouse wheel downwards, the order becomes:

Green, Red, Blue, Yellow, Orange, Purple.

If the player then moves and positions the mouse cursor over the orange player's graphic, and clicks the mouse wheel upwards, the order becomes:

Green, Red, Blue, Orange, Yellow, Purple.

**Reply by zenseii** [reply]
Generally your questions about the implementation are answered by looking at the multiplayer player (:MP) selection menu. **Note that** you should write something for every player's name to actually perceive the change that happens because if you leave the players with the standard colors as names you won't notice anything happening - this is something that should be addressed so that the player gets some kind of feedback telling him that he has changed something.

> How to correctly interpret the sequence with the left and right buttons? The user can left-click on a player for no particular reason, and then right-click on another player (without intending to swap them) because the right-click also displays an "Opponents" information window.

The displaying of the "Opponents" dialog would be disabled once you've clicked on any player and the red outline/square appears around that player.
>  
> And also, how to differentiate this from player selection. Let's say that in the scenario window, the Blue player is selected by default. !
> Now, if the user wants to play another player (another color), he left-clicks for example on the orange player's graphic and this one is selected (as human player).

Basically you would have to left-click two times to change the human players position. This is how it works in MP menu and I believe the cost of having to press once more is outweighed by the benefits of having this simple implementation. In many ways this implementation is quicker because you would not have to go step-by-step in order to swap players on each end of the order.

> So the user should make a second click on the same graphic and a selection frame should be displayed around the player's graphic to indicate to the user that the interface has switched to a player order change mode and that the next click on another player's graphic will swap their positions in the play order. The problem is that the first click on a player's graphic will change that player to a human player if he wasn't before (if it was possible to play that color as a human player). And then the user will have to reselect the color he wants to play as a human player. Which is a bit confusing and not very convenient.

Yes this would be fixed with the new implementation and this doesn't happen with the multiplayer player selection, this problem doesn't exist there, but it is good to keep in mind this of course.

> For the mouse wheel operation, the user positions the mouse cursor on a player's graphic without clicking. By moving the mouse wheel downwards, this player is moved one step to the right. By moving the mouse wheel up, this player is moved one step to the left. If the player is placed at the far left or far right (at the edge, without a player on each side), the player is also moved to the other extremity (the players are considered as in a "loop"). As long as the user does not move the mouse cursor outside the graphic of the original player whose position he has just changed, it is always this original player who changes his position if the user continues to use the mouse wheel in one direction or the other. If the user wants to change the position of another player, he should move the mouse cursor over the graphic of this other player first.
> 
> For example, with the Blue player graphic (far left) as the origin, the standard order is:
> 
> Blue, Green, Red, Yellow, Orange, Purple.
> 
> With a click of the mouse wheel downwards, the order becomes:
> 
> Green, Blue, Red, Yellow, Orange, Purple.
> 
> With another click of the mouse wheel downwards, the order becomes:
> 
> Green, Red, Blue, Yellow, Orange, Purple.
> 
> If the player then moves and positions the mouse cursor over the orange player's graphic, and clicks the mouse wheel upwards, the order becomes:
> 
> Green, Red, Blue, Orange, Yellow, Purple.

I'm trying to understand how this would work so you'll excuse me if I still have some questions. Basically you would hold the cursor over the blue player (or whatever player you want to change the place and order of) and then roll the wheel downward, or do you mean to press the mouse wheel button?

Btw talking about buttons on the mouse, what is your opinion on using the extra side buttons that are almost universally found on mice today? Do you think it is okay to be able to map those for certain actions? I'm talking about this in general for fheroes2 and not about this topic specifically.

EDIT: 
Okay so I reread what you wrote and you do indeed mean to scroll the mouse wheel and I understand how it works. The clear benefit of Oleg's way is that if you want to directly swap purple and blue, you would accomplish that with two mouse clicks instead of first scrolling one player to the other player's position, and then scrolling the other player to the first player's position, which would end up being like this using the mouse wheel, if I understood correctly:

Cursor on the origin icon's color in bold:
Green, Red, Blue, Orange, Yellow, **Purple**.
Two upwards scrolls with cursor pointing on Purple:
Green, Red, Blue, **Purple**, Orange, Yellow.
Then move cursor to be above Orange player's icon and scroll the wheel down once:
Green, Red, Blue, Purple , Yellow, **Orange**.

I'm not saying these two modes exclude eachother from existing, they don't, but since there's already code for Oleg's way I think it's much easier to use that, and make the small additions I talked about.

**Reply by LeHerosInconnu** [reply]
Hello @zenseii,

> Generally your questions about the implementation are answered by looking at the multiplayer player (:MP) selection menu. **Note that** you should write something for every player's name to actually perceive the change that happens because if you leave the players with the standard colors as names you won't notice anything happening - this is something that should be addressed so that the player gets some kind of feedback telling him that he has changed something.

In fheroes2 currently, for the multiplayer, it doesn't work like in the original game.
I will open a new issue for that. (https://github.com/ihhub/fheroes2/issues/4721)

After switching the position of the Blue player and the Orange player.
In the original game, the order of play of the player colors has been changed and the Orange player plays first and at the Orange player position on the adventure map.

https://user-images.githubusercontent.com/43583994/143768348-09041b73-18d7-4f97-a930-b7938ba295b3.mp4

In fheroes2, this is not the case.

https://user-images.githubusercontent.com/43583994/143768416-1d24e872-d62c-439c-8f5f-d42dbfd5d950.mp4

> > How to correctly interpret the sequence with the left and right buttons? The user can left-click on a player for no particular reason, and then right-click on another player (without intending to swap them) because the right-click also displays an "Opponents" information window.
> 
> The displaying of the "Opponents" dialog would be disabled once you've clicked on any player and the red outline/square appears around that player.
> 
> > And also, how to differentiate this from player selection. Let's say that in the scenario window, the Blue player is selected by default. !
> > Now, if the user wants to play another player (another color), he left-clicks for example on the orange player's graphic and this one is selected (as human player).
> 
> Basically you would have to left-click two times to change the human players position. This is how it works in MP menu and **I believe the cost of having to press once more is outweighed by the benefits of having this simple implementation.** In many ways this implementation is quicker because you would not have to go step-by-step in order to swap players on each end of the order.

I think it's better to leave the action to change the color of the human player in single player mode to a single click as in the original game.
Indeed there will be many more players who will use this basic action, compared to the marginal number of players who will want to change the order of play of the colors.

> > For the mouse wheel operation, the user positions the mouse cursor on a player's graphic without clicking. By moving the mouse wheel downwards, this player is moved one step to the right. By moving the mouse wheel up, this player is moved one step to the left. If the player is placed at the far left or far right (at the edge, without a player on each side), the player is also moved to the other extremity (the players are considered as in a "loop"). As long as the user does not move the mouse cursor outside the graphic of the original player whose position he has just changed, it is always this original player who changes his position if the user continues to use the mouse wheel in one direction or the other. If the user wants to change the position of another player, he should move the mouse cursor over the graphic of this other player first.
> > For example, with the Blue player graphic (far left) as the origin, the standard order is:
> > Blue, Green, Red, Yellow, Orange, Purple.
> > With a click of the mouse wheel downwards, the order becomes:
> > Green, Blue, Red, Yellow, Orange, Purple.
> > With another click of the mouse wheel downwards, the order becomes:
> > Green, Red, Blue, Yellow, Orange, Purple.
> > If the player then moves and positions the mouse cursor over the orange player's graphic, and clicks the mouse wheel upwards, the order becomes:
> > Green, Red, Blue, Orange, Yellow, Purple.
> 
> I'm trying to understand how this would work so you'll excuse me if I still have some questions. Basically you would hold the cursor over the blue player (or whatever player you want to change the place and order of) and then roll the wheel downward, or do you mean to press the mouse wheel button?
> 
> EDIT: Okay so I reread what you wrote and you do indeed mean to scroll the mouse wheel and I understand how it works. The clear benefit of Oleg's way is that if you want to directly swap purple and blue, you would accomplish that with two mouse clicks instead of first scrolling one player to the other player's position, and then scrolling the other player to the first player's position, which would end up being like this using the mouse wheel, if I understood correctly:
> 
> Cursor on the origin icon's color in bold: Green, Red, Blue, Orange, Yellow, **Purple**. Two upwards scrolls with cursor pointing on Purple: Green, Red, Blue, **Purple**, Orange, Yellow. Then move cursor to be above Orange player's icon and scroll the wheel down once: Green, Red, Blue, Purple , Yellow, **Orange**.

Yes, that's how it should works. :)

> I'm not saying these two modes exclude eachother from existing, they don't, but since there's already code for Oleg's way I think it's much easier to use that, and make the small additions I talked about.

It's faster to swap two players, yes, but what about inserting a player between two others?
For example moving player 1 between players 5 and 6.
That means a lot of clicks.

> Btw talking about buttons on the mouse, what is your opinion on using the extra side buttons that are almost universally found on mice today? Do you think it is okay to be able to map those for certain actions? I'm talking about this in general for fheroes2 and not about this topic specifically.

Generally, nowadays, the basic mouse has a left button, a right button, a mouse wheel and a middle button (by pressing the mouse wheel).
Personally, I have a mouse like that. :)
I think we should stay with this basic mouse as a reference.

Consideration should also be given to alliances between players, e.g., "Blue, Green, Red vs. other(s)".
Also, the scenario creator can include references to player color in the scenario description and for general events and events on the adventure map, as well as for rumors.

**Comment by oleg-derevenetz** [comment]
Hi @zenseii Please note that in multiplayer mode all this switching works in a different way: you can "select" one player and switch all his settings (human/ai, name and starting race) with any other player directly. This is not the case in the single-player game because when I reworked all this (at the time) I tried to replicate the OG behavior in this regard (the only exception is that OG does not allow to carry the initial race in the multiplayer mode while my implementation allows this).

P.S. I reworked all this in PR #3006 that contains a separate ("rollbackable") commit that changes the meaning of those buttons with arrows because the idea of changing order of players without changing colors looked a bit odd to me. BTW, according to `SortPlayers()` in the `game_startgame.cpp` it never worked anyway because color is taken into account when players are sorted in the `Interface::Basic::StartGame()`.

P.P.S. I believe that all those buttons can be safely removed and UI in single player may be reworked to work as in multiplayer mode instead (not by default, but we can "select" player when the shift button is pressed or something like that). This should be relatively easy to implement, so if there are willing persons, they are free to do this :)

**Reply by zenseii** [reply]
I see! Thanks for the thorough explanation Oleg! I never thought it would behave differently in singleplayer compared to multiplayer but I'll try it out there and see how it works.

It seems we have mostly established that the arrow symbols can be removed, but that implementing a new way of changing the order or placememt of a player specifically in single player mode should be done. 

As for using the shift key, I have to reference @LeHerosInconnu above about taking into consideration players who only want, or only can, use the mouse for playing. Yesterday I realized there is moving of windows by drag 'n' drop in the adventure map if you choose the "hidden" interface which lets you mobe around the various windows. 

Maybe this same function could be implemented in a way for moving the player icons around. It probably won't be hard to specify which players should be selectable or not for this too.

I'll see if I try to take a jab at this, but I have some other stuff I would like to wrap up first and also this depends on how much I will have to change about my current PRs. if anyone in the meantime wants to change this then they are free to try for me. 

I'll post here when I decide to work on it, to avoid duplicate work.

**Reply by oleg-derevenetz** [reply]
@zenseii

> I never thought it would behave differently in singleplayer compared to multiplayer but I'll try it out there and see how it works.

Here's how it works in OG in multiplayer mode:

https://user-images.githubusercontent.com/32623900/143682734-de22ec86-e2a8-4ae3-88a8-7a3931af8371.mp4

and here's how it works in fheroes2:

https://user-images.githubusercontent.com/32623900/143682796-f6bace80-6c17-4c1b-ba5f-245ad721ac11.mp4

**Reply by zenseii** [reply]
For some reason I didn't get a notification about your reply, but thanks for sharing these videos. Before seeing your reply I went ahead and checked it myself and it works very well. It would indeed be great to have this in the single player window too.

I am however trying to think of an easy solution for people who only use the mouse to also have this option available. Maybe if you first left click a player icon, and then you have the choice of left-clicking to simply swap places like before, or right-clicking to change the place and order? I think this is the best and simplest solution, and it fits more in with the general way the game's UI currently works.

This solution would be very much analogous to how creature moving and splitting works in the army bar/garrison, where you would also have the possibility to use the shift key.
